tests/devices/test_default_qubit.py

```diff
@@ -1953,7 +1956,40 @@ def test_apply_two_qubit_op_reverse(self, op, method, inverse):
         op = op(wires=[2, 1])
         matrix = op.inv().matrix if inverse else op.matrix
         matrix = matrix.reshape((2, 2, 2, 2))
-        state_out_einsum = np.einsum("abcd,idc->iba", matrix, self.state)
+        state_out_einsum = np.einsum("abcd,idck->ibak", matrix, self.state)
+        assert np.allclose(state_out, state_out_einsum)
+
+    @pytest.mark.parametrize("op, method", three_qubit_ops)
+    def test_apply_three_qubit_op_controls_smaller(self, op, method, inverse):
+        """Test if the application of three qubit operations is correct when both control wires are
+        smaller than the target wire."""
+        state_out = method(self.state, axes=[0, 2, 3])
+        op = op(wires=[0, 2, 3])
```

Comment:
I've started pulling out the wires variables and parameterizing the tests over the wire order, but I'm having trouble elegantly replacing the `einsum` functions with `tensordot` functions. `tensordot` does not provide a way of reordering the noncontracted dimensions, so I'm instead needing to slice along the control axis, apply `tensordot` to each slice, and then reassemble the state. But this seems much too similar to the actual implementation to be a fair test. For example, in `test_apply_single_qubit_op`, if `wires_order = [0]`, then our test state can simply be constructed using `state_out_tensordot = np.tensordot(matrix, self.state)`. But if `wires_order = [1]`, we need to do
```
result0 = np.tensordot(matrix, self.state[0])
result1 = np.tensordot(matrix, self.state[1])
state_out_tensordot = np.stack([result0, result1])
```
which gets even more involved for higher dimensions, and therefore is hard to generalize without basically copying the `_apply_toffoli` implementation. I've tried specifying different axes along which to apply `tensordot`, but the problem again is the ordering of dimensions. `tensordot` is fine for `einsum` operations such as `ab,bijk->aijk`, because the output is the default ordering. But it seems operations such as `ab,ibjk->iajk` are not reproducible in `tensordot` without further specification. And so the general case i.e. making one specification that works for any `wires_order`, will require `if/else` statements inside the testing function, which is not ideal. Do you have any ideas?

Reply:
Sorry this didn't work out. The idea was to make the tests simpler, but this sounds like it just makes the tests more complicated.  

It was more of a suggestion than a need, so I'm going ahead with changing my review to "approve".  Thanks for trying it out anyway 👍

Reply:
Ok, I think that's right the call. The parameterized version of `test_apply_single_qubit_op` would have needed to look something like this:

<!-- language: python3 -->
    num_wires = 4
    state = np.arange(2 ** num_wires)

    ...

    single_qubit_wires_order = [
        [1],
    ]

    ...

    @pytest.mark.parametrize("op, method", single_qubit_ops)
    @pytest.mark.parametrize("wires_order", single_qubit_wires_order)
    def test_apply_single_qubit_op(self, op, method, wires_order, inverse):
        """Test if the application of single qubit operations is correct."""
        state_out = method(self.state.reshape([2] * self.num_wires), axes=wires_order, inverse=inverse)
        op = op(wires=wires_order)
        matrix0 = op.inv().matrix if inverse else op.matrix
        matrix1 = np.kron(np.eye(2 ** wires_order[0]), matrix0)
        matrix2 = np.kron(matrix1, np.eye(2 ** (self.num_wires - wires_order[0] - 1)))
        state_out_dot = np.dot(matrix2, self.state).reshape([2] * self.num_wires)
        assert np.allclose(state_out, state_out_dot)